src/matchers/to-be-visible.tsx

```diff
@@ -0,0 +1,64 @@
+import type { ReactTestInstance } from 'react-test-renderer';
+import { matcherHint } from 'jest-matcher-utils';
+import { StyleSheet } from 'react-native';
+import { getHostParent } from '../helpers/component-tree';
+import { checkHostElement, formatElement } from './utils';
+
+function isVisibleForStyles(element: ReactTestInstance) {
+  const style = element.props.style || {};
+  const { display, opacity } = StyleSheet.flatten(style);
+  return display !== 'none' && opacity !== 0;
+}
+
+function isVisibleForAccessibility(element: ReactTestInstance) {
+  return (
+    !element.props.accessibilityElementsHidden &&
+    element.props.importantForAccessibility !== 'no-hide-descendants' &&
+    !element.props['aria-hidden']
+  );
+}
+
+function isModalVisible(element: ReactTestInstance) {
+  return element.type.toString() !== 'Modal' || element.props.visible !== false;
```

Comment:
Had to check `element.type` here converted to String instead of using the actual `Modal` class (as done in Jest Native). Any clues why @mdjastrzebski? Seems to work with no downside though

Reply:
I think we should skip the Modal code, as it relies heavily on RN internal implementation details.

Reply:
@mdjastrzebski You're saying to remove the Modal check? To me it feels like an edge-case that should be handled, equivalent to the `<details />` check in Jest DOM

Reply:
`jest-native` has an util for checking type of an element. I imported it in my PR #1469. 

https://github.com/testing-library/jest-native/blob/main/src/utils.ts#L68

Reply:
Let's not do it the way @jaworek suggested. This method is flawed, as it relies on RN implementation details which might change. We have some tools for checking host element type, but should limit it to core components like `Text`, `TextInput`, `View`, etc. In case of Modal seems like pretty primitive mock is actually used: https://github.com/facebook/react-native/blob/main/packages/react-native/jest/mockModal.js

Reply:
This is host element tree for modals:

1. `visible={true}`
```
test('Modal render', () => {
  render(
    <Modal>
      <View testID="content" />
    </Modal>
  );

  expect(screen.toJSON()).toMatchInlineSnapshot(`
    <Modal
      hardwareAccelerated={false}
      visible={true}
    >
      <View
        testID="content"
      />
    </Modal>
  `);
});
```

1. `visible={false}`
```
test('Modal render', () => {
  render(
    <Modal visible={false}>
      <View testID="content" />
    </Modal>
  );

  expect(screen.toJSON()).toMatchInlineSnapshot(`
    <Modal
      hardwareAccelerated={false}
      visible={false}
    />
  `);
});

```

Reply:
@mdjastrzebski Ok, makes sense. But I did not understand what action are you suggesting to take. Remove the `isModalVisible()` check and tests?

Reply:
So regarding **content** of Modal we are good, as this won't be rendered at all. However, the `Modal` host element being rendered when modal is not visible seems incorrect. I'll drop a quick PR there to make the modal not render any host elements if it's not visible.

As a workaround we should be to extend [`hostComponentNames`](https://github.com/callstack/react-native-testing-library/blob/cf08c069c2a0b573297c30e1a920523281d48eeb/src/helpers/host-component-names.tsx#L31) feature to detect the Modal host component name, and then use that value to make check in the Jest matcher, similar to the one from legacy Jest Native matcher.

Reply:
Sent PR to RN repo https://github.com/facebook/react-native/pull/39157